Iaco 586 handle jockey events

tomerlevi444 opened the pull request:
- Handle Jockey events.
- Preferences page.
- Unit test framework: Junit4 + Robolectric + PowerMock + Mockito.
- Tests.

Commit:
added JockeyJS functionality in WebviewActivity and RegistrationActivity

Commit:
FIX: javascript command "window.location = <redirect_url>" redirected to <redirect_url> twice.

Commit:
Added preferences activity.

Commit:
removed Powermock+Easymock from build.gradle

Commit:
Added support for JUnit4 + Robolectric + Powermock + Easymock
changed tests directory from 'androidTest' to 'test'

Commit:
Converted tests to JUnit4

Commit:
Added missing configuration for JUnit4 tests on Android Studio.

Commit:
Switched from Easymock to Mockito.

Commit:
Switched from Easymock to Mockito.

Commit:
Merge branch 'iaco-586-handle-jockey-events' of https://github.com/yuval-netanel/klarna-on-demand-android into iaco-586-handle-jockey-events

Conflicts:
	klarna-on-demand/src/test/java/com/klarna/ondemand/ContextTest.java

Commit:
removed unnecessary configuration to use emulated sdk 18 for JUnit tests

Commit:
Removed Powermock usage in tests.

Commit:
WebViewActivity returns homeButtonResultCode() when home button is pressed.
Dismiss alert appears only on RegistrationActivity.

Commit:
Added tests for Registration and Preferences activities.

Commit:
removed dependency on Powermock for stubbing Context static class.

Commit:
merge with Master

Commit:
Upgraded to JUnit 4.12 and added Tests.

yuval-netanel commented on example/src/main/java/com/klarna/example/MainActivity.java:
```diff
@@ -19,23 +20,69 @@ protected void onCreate(Bundle savedInstanceState) {
+    }
+
+    private void saveUserToken(String token) {
+        SharedPreferences preferences = getPreferences(0);
```
Context.MODE_PRIVATE instead of 0

tomerlevi444 replied:
Done

yuval-netanel commented on example/src/main/java/com/klarna/example/MainActivity.java:
```diff
@@ -19,23 +20,69 @@ protected void onCreate(Bundle savedInstanceState) {
+                    String token = data.getStringExtra(RegistrationActivity.EXTRA_USER_TOKEN);
+                    saveUserToken(token);
+                    findViewById(R.id.registerButton).setVisibility(View.INVISIBLE);
+                    findViewById(R.id.preferencesButton).setVisibility(View.VISIBLE);
```
This and initializeUIElements() have the same logic, and can be in one toggleButton function

grozen replied:
+1

yuval-netanel commented on import-summary.txt:
```diff
@@ -0,0 +1,60 @@
+ECLIPSE ANDROID PROJECT IMPORT SUMMARY
```
Was ist das?

yuval-netanel commented on jockeyJSAndroid/build.gradle:
```diff
@@ -0,0 +1,22 @@
+apply plugin: 'com.android.library'
```
@grozen we should make sure that jockeyJSAndroid is in our aar

grozen replied:
As long as we don't end up with two separate aars after running `gradle build` then we should be ok. Won't you check whether or not that is the case, @tomerlevi444?

yuval-netanel commented on jockeyJSAndroid/build.gradle:
```diff
@@ -0,0 +1,22 @@
+}
+
+dependencies {
+    compile 'com.google.code.gson:gson:2.2.4'
```
@grozen is gson going to be in our pom.xml or it's going to be included in the arr?

grozen replied:
I **think** that as long as we define the dependency we are good. Also, if every aar/jar had to bundle it's own dependencies with itself things would quickly get out of hand, so I suspect it's unnecessary.

yuval-netanel commented on klarna-on-demand/build.gradle:
```diff
@@ -29,15 +17,24 @@ android {
-        exclude group : 'org.hamcrest'
-        exclude module: 'objenesis'
+    sourceSets {
+        androidTest.setRoot('src/test')
```
You should update the test location in the readme

tomerlevi444 replied:
Done
Thanks!

yuval-netanel commented on klarna-on-demand/src/main/java/com/klarna/ondemand/PreferencesActivity.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    @Override
+    protected void handleUserReadyEventWithPayload(Map<Object, Object> payload) {
```
I would call it  handleUserReadyEvent, not like IOS, WDYT?

grozen replied:
I second that motion. Saying "WithPayload" when you have a payload argument is somewhat redundant.

tomerlevi444 replied:
Changed.

yuval-netanel commented on klarna-on-demand/src/main/java/com/klarna/ondemand/PreferencesActivity.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    @Override
+    public void onBackPressed() {
```
protected

tomerlevi444 replied:
But it's public at the Activity base class.

yuval-netanel commented on klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationActivity.java:
```diff
@@ -1,12 +1,61 @@
  */
 public class RegistrationActivity extends WebViewActivity {
 
+    public static final String EXTRA_USER_TOKEN = "userToken";
```
You can put it in WebViewActivity, because both registration and preference use it

yuval-netanel commented on klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationActivity.java:
```diff
@@ -1,12 +1,61 @@
     }
+
+    @Override
+    protected void handleUserReadyEventWithPayload(Map<Object, Object> payload) {
```
handleUserReadyEvent

tomerlevi444 replied:
Done

yuval-netanel commented on klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationActivity.java:
```diff
@@ -1,12 +1,61 @@
+
+    @Override
+    protected void handleUserReadyEventWithPayload(Map<Object, Object> payload) {
+        String token = (String) payload.get("userToken");
```
EXTRA_USER_TOKEN

yuval-netanel commented on klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationActivity.java:
```diff
@@ -1,12 +1,61 @@
+    }
+
+    @Override
+    public void onBackPressed() {
```
protected

yuval-netanel commented on klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationActivity.java:
```diff
@@ -1,12 +1,61 @@
+        builder.setMessage(R.string.DISMISS_ALERT_MESSAGE);
+        builder.setCancelable(true);
+
+        builder.setPositiveButton("Yes", new DialogInterface.OnClickListener() {
```
Need translation

tomerlevi444 replied:
Done

yuval-netanel commented on klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationActivity.java:
```diff
@@ -1,12 +1,61 @@
+            }
+        });
+
+        builder.setNegativeButton("No", new DialogInterface.OnClickListener() {
```
Need translation

tomerlevi444 replied:
Done

yuval-netanel commented on klarna-on-demand/src/main/java/com/klarna/ondemand/WebViewActivity.java:
```diff
@@ -41,22 +52,37 @@ public boolean onOptionsItemSelected(MenuItem item) {
-    public void onBackPressed() {
-        showDismissAlert();
+    protected void onDestroy() {
+        jockey.off("userReady");
```
Use const for both jokey events

tomerlevi444 replied:
Done

yuval-netanel commented on klarna-on-demand/src/test/java/com/klarna/ondemand/ContextTest.java:
```diff
@@ -0,0 +1,31 @@
+public class ContextTest {
+
+    @Rule
+    public ExpectedException thrown = ExpectedException.none();
```
You should change the methods name according to the new convention

yuval-netanel commented on klarna-on-demand/src/test/java/com/klarna/ondemand/ContextTest.java:
```diff
@@ -0,0 +1,31 @@
+    public ExpectedException thrown = ExpectedException.none();
+
+    @Test
+    public void setApiKey() {
```
You should change the methods name according to the new convention

tomerlevi444 replied:
Done

yuval-netanel commented on klarna-on-demand/src/test/java/com/klarna/ondemand/RegistrationActivityTest.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    @Test
+    public void handleUserReadyEventWithPayload_ShouldCallFinishWithRESULTOK_WhenATokenIsReceived() {
```
//region .handleUserErrorEvent

yuval-netanel replied:
handleUserReadyEventWithPayload_ShouldCallFinishWithResultOk_WhenATokenIsReceived

tomerlevi444 replied:
Done

yuval-netanel commented on klarna-on-demand/src/test/java/com/klarna/ondemand/RegistrationActivityTest.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    @Test
+    public void handleUserErrorEvent_ShouldCallFinishWithRESULTERROR() {
```
handleUserErrorEvent_ShouldCallFinishWithResultError

tomerlevi444 replied:
Done

yuval-netanel commented on klarna-on-demand/src/test/java/com/klarna/ondemand/RegistrationActivityTest.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    @Test
+    public void homeButtonPress_ShouldCallFinishWithRESULTCANCELED() {
```
homeButtonPress_ShouldCallFinishWithResultCanceled

tomerlevi444 replied:
Done

grozen commented on example/src/main/java/com/klarna/example/MainActivity.java:
```diff
@@ -19,23 +20,69 @@ protected void onCreate(Bundle savedInstanceState) {
+                    findViewById(R.id.preferencesButton).setVisibility(View.VISIBLE);
+                    break;
+                case RegistrationActivity.RESULT_CANCELED:
+                case RegistrationActivity.RESULT_ERROR:
```
In the iOS demo we had a comment along the lines of 

``` objective-c
// You may also want to convey this failure to your user.
```

when things fail. We should add one here as well.

tomerlevi444 replied:
Done

grozen commented on example/src/main/java/com/klarna/example/MainActivity.java:
```diff
@@ -19,23 +20,69 @@ protected void onCreate(Bundle savedInstanceState) {
+        else if(requestCode == PREFERENCES_REQUEST_CODE) {
+            switch(resultCode) {
+                case PreferencesActivity.RESULT_OK:
+                case PreferencesActivity.RESULT_ERROR:
```
Same deal here.

tomerlevi444 replied:
Done

grozen commented on example/src/main/java/com/klarna/example/MainActivity.java:
```diff
@@ -19,23 +20,69 @@ protected void onCreate(Bundle savedInstanceState) {
+    }
+
+    private String getUserToken() {
+        SharedPreferences preferences = getPreferences(0);
```
In case you didn't change this one as well, do.

tomerlevi444 replied:
Done

grozen commented on jockeyJSAndroid/src/main/AndroidManifest.xml:
```diff
@@ -0,0 +1,10 @@
+<manifest xmlns:android="http://schemas.android.com/apk/res/android"
```
Having this as an additional project that generates an individual aar (I also suspect this should end up as a jar and not an aar) seems like a little bit of an overkill.
This isn't really an individual package that we encourage people to use outside of our SDK, and I suspect this will also cause some grief when utilizing the SDK from maven.
The only reason to do this is if it's considered "proper", and I suspect @MaTriXy is just the man to answer that question.

tomerlevi444 replied:
I've integrated the JockeyJSAndroid project into our project (by migrating the 'src/java' library).
Now it all compiles into the same `.aar` file.

grozen commented on klarna-on-demand/build.gradle:
```diff
@@ -1,18 +1,6 @@
 apply plugin: 'com.android.library'
 
 android {
-    packagingOptions {
```
Why drop this?

tomerlevi444 replied:
Apparently it is not needed after the changes on build.gradle's dependencies section.

grozen commented on klarna-on-demand/build.gradle:
```diff
@@ -29,15 +17,24 @@ android {
+}
 
-    androidTestCompile 'org.easymock:easymock:3.3'
+apply plugin: 'android-unit-test'
```
With all these changes in the gradle, it would be prudent to check if the [publishing instructions](https://confluence.internal.machines/display/teamindex/Deploying+to+Maven) still work.

tomerlevi444 replied:
I've run `./gradlew uploadArchives` and the build passed. Is it enough?

grozen replied:
@tomerlevi444 almost. Be sure to go to the site listed in the deployment guide, look at the recently uploaded package and make sure nothing is missing inside. If the guide isn't clear enough, I'll gladly walk you through it tomorrow.

grozen commented on klarna-on-demand/src/test/java/com/klarna/ondemand/PreferencesActivityTest.java:
```diff
@@ -0,0 +1,94 @@
+
+
+@RunWith(RobolectricTestRunner.class)
+@Config(emulateSdk = 18)
```
Why 18?

tomerlevi444 replied:
Robolectric supports android SDK <= 18

grozen commented on klarna-on-demand/src/test/java/com/klarna/ondemand/PreferencesActivityTest.java:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    @Test
+    public void handleUserErrorEvent_ShouldCallFinishWithRESULTERROR() {
```
There are obviously no hard rules for this, but I think I would prefer this method be called `handleUserErrorEvent_shouldCallFinishWith_RESULT_ERROR`. Your call.

This obviously applies to the other test names as well.

tomerlevi444 replied:
@grozen What about `handleUserErrorEvent_shouldCallFinishWithResultError` ?

grozen replied:
I could go for that.

On Wed, Jan 14, 2015 at 5:36 PM, tomerlevi444 notifications@github.com
wrote:

> In
> klarna-on-demand/src/test/java/com/klarna/ondemand/PreferencesActivityTest.java
> https://github.com/klarna/klarna-on-demand-android/pull/2#discussion-diff-22943907
> :
> 
> > +
> > -    private PreferencesActivity preferencesActivity;
> > -    private ActivityController<PreferencesActivity> preferencesActivityController;
> >   +
> > -    @Before
> > -    public void beforeEach() {
> > -        PowerMockito.mockStatic(Context.class);
> > -        Mockito.when(Context.getApiKey()).thenReturn("test_skadoo");
> >   +
> > -        Intent intent = new Intent(Intent.ACTION_VIEW);
> > -        intent.putExtra(preferencesActivity.EXTRA_USER_TOKEN, "my_token");
> > -        preferencesActivityController = Robolectric.buildActivity(PreferencesActivity.class).withIntent(intent).create();
> > -    }
> >   +
> > -    @Test
> > -    public void handleUserErrorEvent_ShouldCallFinishWithRESULTERROR() {
> 
> @grozen https://github.com/grozen What about
> handleUserErrorEvent_shouldCallFinishWithResultError ?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/klarna/klarna-on-demand-android/pull/2/files#r22943907
> .

## 

Die Aubergine ist purpurrot

grozen commented on klarna-on-demand/src/test/java/com/klarna/ondemand/RegistrationActivityTest.java:
```diff
@@ -0,0 +1,84 @@
+        Mockito.verify(registrationActivity).setResult(RegistrationActivity.RESULT_CANCELED);
+        Mockito.verify(registrationActivity).finish();
+    }
+}
```
Why doesn't this have a back button test?

tomerlevi444 replied:
Added a test that checks the showDismissAlert is called on back button press.

grozen commented on klarna-on-demand/src/test/java/com/klarna/ondemand/UrlHelperTest.java:
```diff
@@ -0,0 +1,82 @@
+
+    @Test
+    public void registrationUrl_ShouldReturnProductionUrl_WhenTokenIsForProduction() {
+        PowerMockito.mockStatic(Context.class);
```
What happens when you mocked something that's already mocked?
Do mocks clear each test?

tomerlevi444 replied:
When you mock something that's already mocked, it replaces the return values. 
Yes, mocks clear after each test (both static and non-static mocks).
I've also revalidated this.

grozen commented on klarna-on-demand/src/test/java/com/klarna/ondemand/UrlHelperTest.java:
```diff
@@ -0,0 +1,82 @@
+    public void registrationUrl_ShouldReturnUrlWithSwedishLocale_WhenLocaleIsSwedish() {
+        Locale.setDefault(new Locale("sv"));
+
+        Assert.assertTrue(UrlHelper.registrationUrl().contains("locale=sv"));
```
As I said before in iOS - this is a great test, but it doesn't assure us that this is not something that happen even when the locale isn't 'sv'.

yuval-netanel replied:
+1

tomerlevi444 replied:
Done

grozen commented:
This some terrific work!

tomerlevi444 committed:
Delete import-summary.txt

Commit:
Review changes

Commit:
Merge branch 'iaco-586-handle-jockey-events' of https://github.com/klarna/klarna-on-demand-android into iaco-586-handle-jockey-events

Commit:
Merge branch 'master' into iaco-586-handle-jockey-events

Commit:
concatenate string using '+' instead of interpolation.

tomerlevi444 commented:
Some additional notes:
- Added static imports in the test classes (so instead of writing `Mockito.when`, you'll write `when`), which is much neater.
- Added documentation for PreferencesActivity.EXTRA_USER_TOKEN and RegistrationActivity.EXTRA_USER_TOKEN.
- Marked Context class as final.

yuval-netanel commented:
Nice job! approved

Commit:
store findViewById's result in a field.

Commit:
changed initializeUIElements to updateUIElements

Commit:
semantic change.

grozen commented on klarna-on-demand/src/test/java/com/klarna/ondemand/UrlHelperTest.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    @Test
+    public void registrationUrl_ShouldReturnUrlWithSwedishLocale_WhenLocaleIsSwedish() {
```
The test names don't make sense anymore, now that we are no longer talking about Swedish.
Also, the inherent problem remains. When the locale is set to "Swedish" we test to see that the generated URLs reflect it, but we have no test that verifies that if the locale **is not** "Swedish", the URL **does not** contain it.

grozen commented:
Approved(!)

tomerlevi444 committed:
Update README.md

tomerlevi444 committed:
Update README.md

tomerlevi444 commented:
@grozen @yuval-netanel 
I've added instructions on setting up a JUnit4 environment on Android Studio. please have a look at: https://github.com/klarna/klarna-on-demand-android/blob/master/doc/android_studio_setup.md

yuval-netanel commented on README.md:
```diff
@@ -53,7 +53,9 @@ As per [running the application](#running_the_application).
-Pull requests must include tests for their proposed fixes/functionality. We use [JUnit](http://junit.org/) to write our tests and you can find the tests themselves [here](klarna-on-demand/src/androidTest/java/com/klarna/ondemand).
+Pull requests must include tests for their proposed fixes/functionality. We use [JUnit](http://junit.org/) to write our tests and you can find the tests themselves [here](klarna-on-demand/src/test/java/com/klarna/ondemand).
+
+If you're using Android Studio, use the [following guide](https://github.com/klarna/klarna-on-demand-android/blob/master/doc/android_studio_setup.md) in order to setup a JUnit testing environment.
```
Use the following guide in order to run our JUnit 4 tests on Android studio?

grozen replied:
you should use relative links, as in `[following guide](doc/android_studio_setup.md)`.

Seems I've also sinned, and you can probably change the link on line 56 to `src/test/java/com/klarna/ondemand`.

Finally, I approve of Yuval's notion, but drop the "on Android studio" at the end as we already says this only applies to Android Studio.

tomerlevi444 committed:
Delete klarna-on-demand-android-tmp.iml

Commit:
Changed UrlHelper Tests' name.

Commit:
merge with Master

Commit:
moved JockeyJSAndroid project into klarna-on-demand.

Commit:
added compiled iml file.

Commit:
modify README.md and android_studio_setup.md

tomerlevi444 committed:
Update android_studio_setup.md

Commit:
before running the tests, copy resources from 'res/test' to 'res'

Commit:
merge with Master
